Lib/functools.py

```diff
@@ -273,33 +273,96 @@ def reduce(function, sequence, initial=_initial_missing):
 ### partial() argument application
 ################################################################################
 
+
+class PlaceholderType:
+    """PlaceholderType()
+    --
+
+    The type of the Placeholder singleton.
+    Used as a placeholder for partial arguments.
+    """
+    _instance = None
+    def __new__(cls):
+        if cls._instance is None:
+            cls._instance = object.__new__(cls)
+        return cls._instance
+
+    def __repr__(self):
+        return 'Placeholder'
+
+    def __bool__(self):
+        raise TypeError("Placeholder should not be used in a boolean context")
+
+
+Placeholder = PlaceholderType()
```

Comment:
```suggestion
class _PlaceholderType:
    """The type of the Placeholder singleton.

    Used as a placeholder for partial arguments.
    """

    def __repr__(self):
        return 'Placeholder'

    def __bool__(self):
        raise TypeError("Placeholder should not be used in a boolean context")

Placeholder = _PlaceholderType()
```

No need for a singleton pattern like that. You just need to create a single instance of the `_PlaceholderType` class. This is similar to what is done in `dataclasses` for `MISSING` and other sentinel objects.

You would also update the C implementation if you decide to use `PLACEHOLDER` and `_PLACEHOLDER_TYPE` (private).

Reply:
But this I think is best to be left as it is to be in sync with C.